Add url_to() URL helper function

ImMaax opened the pull request:
This PR adds a URL helper function called `url_to()`.

**Description**
I added a URL helper called `url_to()`. It creates absolute URLs based on controllers. I'm commonly using this function in basically all of my CodeIgniter projects. Here is an example:

```php
service('routes')->get('/welcome', 'Pages::welcome');
echo url_to('Pages::welcome'); # => http://example.com/welcome

service('routes')->get('/page/(:segment)', 'Pages::view/$1');
echo url_to('Pages::view', 'home'); # => http://example.com/page/home
```

It is meant to be used inside of views to generate links. The main advantage of this function is that you can change your routes later on, but not having to change your links across all your views.

This is the first time I'm contributing code to CodeIgniter, I hope I didn't miss something.

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

MGatner commented on system/Helpers/url_helper.php:
```diff
@@ -634,3 +634,27 @@ function mb_url_title(string $str, string $separator = '-', bool $lowercase = fa
+			$controller = service('routes')->getDefaultNamespace() . $controller;
+		}
+
+		return base_url(route_to($controller, ...$args));
```
I *think* this needs to be `site_url` to account for subdirectories - can confirm on desktop later.

ImMaax replied:
If I change `base_url` to `site_url`, it will prepend `index.php` (so `http://example.com/xyz` would become `http://example.com/index.php/xyz`). Otherwise, it's still fine, so changed that. Will commit soon.

MGatner replied:
The `index.php` should appear or not depending on the user's app setting, which I think is good. It could be another test scenario though, to make sure `url_to()` is respecting `config('App')->indexPage`.

ImMaax replied:
No matter what I try, doesn't seem to work, if I try to respect that indexPage setting

ImMaax commented:
> This is awesome! What a great addition, thank you for the idea and PR. See my note on the URL function of choice. I also would like to see a few more tests, maybe a provider would be in order? Things like passing an empty string and `/`, and testing that names routes work as expected. If you can add those that would be great, I'm excited for this one - it will also be very handy with `FeatureResponse::getRedirectUrl()`.

So first off, thank you! For both the feedback and the appreciation. I changed `base_url` to `site_url` and added tests for empty strings as well as passing a slash. What exactly are providers and named routes, though? 😬

@MGatner

MGatner commented:
Thanks for that! So quick too. A data provider is just a way to specify a bunch of test criteria and results for a single test. The way you are doing it is fine, but you could also provide an array of 'route' => 'url' and run it through the same test. Search for `@dataProvider` amping the tests for some examples. 

Named routes are a way of providing an overriding name for a route: https://codeigniter4.github.io/CodeIgniter4/incoming/routing.html#using-named-routes

I think the Router call is handling this for you, it's just hard for me to tel on mobile.

ImMaax commented:
> Thanks for that! So quick too. A data provider is just a way to specify a bunch of test criteria and results for a single test. The way you are doing it is fine, but you could also provide an array of 'route' => 'url' and run it through the same test. Search for `@dataProvider` amping the tests for some examples.
> 
> Named routes are a way of providing an overriding name for a route: https://codeigniter4.github.io/CodeIgniter4/incoming/routing.html#using-named-routes
> 
> I think the Router call is handling this for you, it's just hard for me to tel on mobile.

TIL about named routes in CI and data providers in PHPUnit, thanks for telling me about them! `url_to` is now compatible with named routes and I reworked the tests for it a little. Hope this fits now :)

ImMaax commented:
> Welcome! This is looking great. I can't wait for your next PRs 🤗
> 
> I should have asked for this before, but how about some tests (and docs) for failure handling? Like what happens if I call `url_to('Foo::bar')` and `Foo` doesn't exist, or `bar` isn't a method?
> 
> Your data provider is great, and if you wanted you could expand it to all your tests. Make your toes have two values (class, URL) and your test take two parameters and then you can test them all in one go.

On failure, the function will just return the base URL of `site_url()`, so for example `http://example.com/index.php`. Should this be changed or is it fine as is? I'll also rework the tests once again.

MGatner commented:
IMO a failure should not return a valid URL, because that would imply that there is an existing route for the query. I would return either `null` or even throw an exemption. Given that this should almost always be a developer's internal use I'd be in favor of throwing. Something like:
```
if (! $route = route_to($controller, ...$args))
{
    throw new \CodeIgniter\Router\Exceptions\RouterException(lang('HTTP.invalidRoute', $controller));
}

return site_url($route);
```

People can always catch it if they don't care, but I would want to know if I had typoed my controller name or something instead of providing links to base. Might be a good time for @michalsn or @samsonasik or someone else to give an opinion as well.

ImMaax commented:
> IMO a failure should not return a valid URL, because that would imply that there is an existing route for the query. I would return either `null` or even throw an exemption. Given that this should almost always be a developer's internal use I'd be in favor of throwing. Something like:
> 
> ```
> if (! $route = route_to($controller, ...$args))
> {
>     throw new \CodeIgniter\Router\Exceptions\RouterException(lang('HTTP.invalidRoute', $controller));
> }
> 
> return site_url($route);
> ```
> 
> People can always catch it if they don't care, but I would want to know if I had typoed my controller name or something instead of providing links to base. Might be a good time for @michalsn or @samsonasik or someone else to give an opinion as well.

Yeah, a failure should not do that. However, that's the default behavior of doing `site_url(route_to())`, so I thought it might fit in. But you're right. Changed it, exception is now thrown.

samsonasik commented on system/Helpers/url_helper.php:
```diff
@@ -634,3 +634,37 @@ function mb_url_title(string $str, string $separator = '-', bool $lowercase = fa
+	{
+		if (! $route = route_to($controller, ...$args))
+		{
+			throw new \CodeIgniter\Router\Exceptions\RouterException(lang('HTTP.invalidRoute', [$controller]));
```
`HTTP.invalidRoute` seems for redirect response, probably use `HTTP.controllerNotFound` with uses explode `::` from controller and method variable definition.

ImMaax replied:
Will change that, thank you

michalsn commented:
Personally, I would rather follow the behavior present in the `route_to()` function and return `false` on failure. Throwing an exception is fine for development mode but not really for production (we basically get a white page with error). It's a helper function - nobody will probably bother to catch en exception in a view or every controller (besides it wouldn't be very handy). But it's just my preference.

MGatner commented:
I'm fine with it either way. Generally I would like to see the framework throw more exceptions when things go wrong - throwing and catching is the best way to deal with problems and in cases like this amounts to mostly the same thing as gathering the result and then testing for `false`. I think you're right that helper functions are maybe the exception (haha) to this rule since they are designed for quick-and-easy access to existing class methods, but it still comes down to whether you want a "Whoops" screen or some malformed link on your production page if you've screwed up your code. Difference is, the Whoops will be logged and reported whereas the broken link will need to be noticed.

ImMaax commented:
> I'm fine with it either way. Generally I would like to see the framework throw more exceptions when things go wrong - throwing and catching is the best way to deal with problems and in cases like this amounts to mostly the same thing as gathering the result and then testing for `false`. I think you're right that helper functions are maybe the exception (haha) to this rule since they are designed for quick-and-easy access to existing class methods, but it still comes down to whether you want a "Whoops" screen or some malformed link on your production page if you've screwed up your code. Difference is, the Whoops will be logged and reported whereas the broken link will need to be noticed.

I personally feel like throwing is the better choice in that case and totally agree with you.

paulbalandan commented on user_guide_src/source/helpers/url_helper.rst:
```diff
@@ -360,3 +360,25 @@ The following functions are available:
 
         $url = prep_url('example.com');
+
+.. php:function:: url_to([$controller[, ...$args]])
```
Based on the function signature, `$controller` is not an optional argument, so probably you can remove the square brackets around it.

ImMaax replied:
Oh yea, that was a mistake, fixing that!

paulbalandan commented on tests/system/Helpers/URLHelperTest.php:
```diff
@@ -1250,4 +1250,64 @@ public function testBasedWithOtherIndex()
+
+	public function urlToProvider()
+	{
+		if (config('App')->indexPage != '') {
```
The opening braces should be on its own line.

ImMaax replied:
Fixed it

paulbalandan commented on system/Helpers/url_helper.php:
```diff
@@ -634,3 +634,37 @@ function mb_url_title(string $str, string $separator = '-', bool $lowercase = fa
+
+		if (! $route = route_to($controller, ...$args))
+		{
+			throw new \CodeIgniter\Router\Exceptions\RouterException(lang('HTTP.invalidRoute', [$controller]));
```
I think you have missed doing @samsonasik comment here.

michalsn commented:
I think we need some changes here. We have to make sure that this new helper will be able to support not only [reverse routing](https://codeigniter.com/user_guide/incoming/routing.html#reverse-routing) but also [named routes](https://codeigniter.com/user_guide/incoming/routing.html#using-named-routes). I'm 100% sure people will expect that because it's basically the same feature.

ImMaax commented:
> I think we need some changes here. We have to make sure that this new helper will be able to support not only [reverse routing](https://codeigniter.com/user_guide/incoming/routing.html#reverse-routing) but also [named routes](https://codeigniter.com/user_guide/incoming/routing.html#using-named-routes). I'm 100% sure people will expect that because it's basically the same feature.

Uhh, it does support named routes 😁

michalsn commented:
Looking at the code, I'm pretty sure it does not.

ImMaax commented:
> Looking at the code, I'm pretty sure it does not.

I'm pretty sure it does. Just checked it:

```php
service('routes')->add('/test', 'Test::index', ['as' => 'main']);
echo url_to('main'); # => http://localhost:8080/index.php/test
```

Am I missing something?

michalsn commented:
Yes, try something that is not defined.

ImMaax commented:
> Yes, try something that is not defined.

Yup, you're right, in case of failure, the string is expected to be a class::method notation, changed that.

michalsn commented on system/Helpers/url_helper.php:
```diff
@@ -634,3 +634,46 @@ function mb_url_title(string $str, string $separator = '-', bool $lowercase = fa
+	 */
+	function url_to(string $controller, ...$args): string
+	{
+		if (strpos($controller, '::') !== false)
+		{
+			$class = explode('::', $controller)[0];
+
+			if (! class_exists($class))
+			{
+				$controller = service('routes')->getDefaultNamespace() . $controller;
+			}
+		}
```
We should run this code only if `route_to()` fail. In other case, it's just a waste of time.

MGatner replied:
@ImMaax Looks like we're almost there! Could you move this around and run tests again?

ImMaax replied:
@MGatner sure, sure. I'm currently busy, but will work on this later today!

michalsn commented on system/Helpers/url_helper.php:
```diff
@@ -634,3 +634,51 @@ function mb_url_title(string $str, string $separator = '-', bool $lowercase = fa
+				}
+			}
+
+			if ($route = route_to($controller, ...$args))
+			{
+				return site_url($route);
+			}
```
Almost there. These lines are unnecessary.

ImMaax replied:
@michalsn Sorry, but I don't quite get how they're unneccesary. What should it look like instead?

michalsn replied:
We've already tried using `route_to()` on line 653 - so we know that will fail. Adding the default namespace will not change this.

ImMaax replied:
So now, I've got this:

```php
	function url_to(string $controller, ...$args): string
	{
		if (! $route = route_to($controller, ...$args))
		{
			$explode = explode('::', $controller);

			if (isset($explode[1]))
			{
				throw new \CodeIgniter\Router\Exceptions\RouterException(lang('HTTP.controllerNotFound', [$explode[0], $explode[1]]));
			}
			else
			{
				throw new \CodeIgniter\Router\Exceptions\RouterException(lang('HTTP.invalidRoute', [$controller]));
			}
		}

		return site_url($route);
	}
```

Works well with named routes and if you pass in a namespaced controller, but if I pass in a non-namespaced controller, it will throw an exception. That's why I added the default namespace. Am I still overlooking something?

michalsn replied:
If we want to add the ability to omit the default namespace, this change should happen in the `route_to()` function. I just want to keep the consistency between functions. Of course, this is only my opinion.

ImMaax replied:
Alright, this makes perfect sense and consistency is important to me too. However, I'd like to know what others think about this (@MGatner @samsonasik @paulbalandan just to mention who's been involved in this thread). I'm perfectly fine with both having it in there and not having it in there.

samsonasik commented on system/Helpers/url_helper.php:
```diff
@@ -634,3 +634,51 @@ function mb_url_title(string $str, string $separator = '-', bool $lowercase = fa
+			{
+				throw new \CodeIgniter\Router\Exceptions\RouterException(lang('HTTP.controllerNotFound', [$explode[0], $explode[1]]));
+			}
+			else
```
no need else as already throw early.

paulbalandan replied:
Yes. Useless else here.

michalsn replied:
@ImMaax It would be great if you could fix this one too, but if not then I will merge it tomorrow.

ImMaax replied:
@michalsn Removed the if/else and explode :)

michalsn replied:
@ImMaax We have a little misunderstanding :) Everything was good except this "else", so:
```php
			if (isset($explode[1]))
			{
				throw new \CodeIgniter\Router\Exceptions\RouterException(lang('HTTP.controllerNotFound', [$explode[0], $explode[1]]));
			}

			throw new \CodeIgniter\Router\Exceptions\RouterException(lang('HTTP.invalidRoute', [$controller]));

```

ImMaax replied:
Ohh, haha, alright, will change this real quick

michalsn commented:
@ImMaax Sorry it took so long. We appreciate your contribution!

ImMaax commented:
> @ImMaax Sorry it took so long. We appreciate your contribution!

No problem! Really glad I could help! Looking forward to more contributions! :)
